src/toolkits/object_detection/one_shot_object_detection/one_shot_object_detector.cpp

```diff
@@ -133,85 +134,100 @@ gl_sframe augment_data(const gl_sframe &data,
-      flex_image flex_background = background_ft.to<flex_image>();
-      if (!(flex_background.is_decoded())) {
-        decode_image_inplace(flex_background);
+      for (const auto& row: data.range_iterator()) {
+        // go through all the starter images and create augmented images for
+        // all starter images and the respective chunk of background images
+        flex_image object = row[image_column_index].to<flex_image>();
+        std::string label = row[target_column_index].to<flex_string>();
+        if (!(object.is_decoded())) {
+          decode_image_inplace(object);
+        }
+        flex_image flex_background = background_ft.to<flex_image>();
+        if (!(flex_background.is_decoded())) {
+          decode_image_inplace(flex_background);
+        }
+        size_t background_width = flex_background.m_width;
+        size_t background_height = flex_background.m_height;
+        size_t background_channels = flex_background.m_channels;
+
+        ParameterSampler parameter_sampler = ParameterSampler(
+                                                background_width,
+                                                background_height,
+                                                (background_width-object.m_width)/2,
+                                                (background_height-object.m_height)/2);
+
+        // construct annotation dictionary from parameters
+        flex_dict annotation = build_annotation(parameter_sampler, label,
+                                                object.m_width, object.m_height,
+                                                seed+row_number);
+
+        DASSERT_TRUE(flex_background.get_image_data() != nullptr);
+        DASSERT_TRUE(object.is_decoded());
+        DASSERT_TRUE(flex_background.is_decoded());
+
+        boost::gil::rgba8_image_t::view_t starter_image_view = create_starter_image_view(object);
+
+        boost::gil::rgb8_image_t::view_t background_view = interleaved_view(
+          background_width,
+          background_height,
+          (boost::gil::rgb8_pixel_t*)(flex_background.get_image_data()),
+          background_channels * background_width // row length in bytes
+          );
+        flex_image synthetic_image = create_synthetic_image(starter_image_view, background_view, parameter_sampler, object);
+        encode_image_inplace(synthetic_image);
+        // write the synthetically generated image and the constructed
+        // annotation to output SFrame.
+        output_writer.write({
+          synthetic_image,
+          annotation
+        }, segment_id);
+        augmented_counter++;
+        if (verbose) {
+          std::ostringstream d;
+          // For pretty printing, floor percent done
+          // resolution to the nearest .25% interval.  Do this by multiplying by
+          // 400, then do integer division by the total size, then float divide
+          // by 4.0
+          int augmented_rows_completed = augmented_counter;
```

Comment:
You should not read the atomic again after you increment above; another thread may already have changed it.

Touch the atomic once. I also strongly recommend using the more verbose API in std::atomic to make it clear that this is an atomic, not a normal integer.

https://en.cppreference.com/w/cpp/atomic/atomic/fetch_add

Reply:
Done, thanks for the reference!